File: tests/test_phase16_notification_consolidation.py
from __future__ import annotations

import unittest
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]


def read_source(path: str) -> str:
    return (ROOT / path).read_text(encoding="utf-8")


class Phase16NotificationConsolidationTests(unittest.TestCase):
    def setUp(self):
        self.app = read_source("app.py")
        self.wsgi = read_source("wsgi.py")
        self.worker = read_source("notification_worker.py")
        self.settings_module = read_source("tvtracker/notifications/backend.py")
        self.engine = read_source("tvtracker/notifications/engine.py")
        self.canonical = read_source("tvtracker/notifications/push_and_movies.py")
        self.runtime = read_source("tvtracker/notifications/runtime.py")
        self.validation = read_source("tvtracker/notifications/push_validation.py")
        self.registry = read_source("tvtracker/migrations/registry.py")
        self.versions = read_source("tvtracker/migrations/versions.py")

    def test_single_canonical_settings_table_in_use(self):
        for source in (
            self.settings_module,
            self.canonical,
            self.runtime,
            self.validation,
            self.engine,
            self.app,
            self.wsgi,
            self.worker,
        ):
            self.assertNotIn(
                "tv_tracker_final_notification_settings",
                source,
                "application code must not read the retired settings table",
            )
        self.assertIn(
            "FROM tv_tracker_notification_settings WHERE singleton_id = 1",
            self.settings_module,
        )
        self.assertIn('"movieReleased": "movie_released"', self.settings_module)
        self.assertIn('"movieReleaseUpdates": "movie_release_updates"', self.settings_module)
        self.assertIn('    "movie_released",', self.settings_module)
        self.assertIn('    "movie_release_updates",', self.settings_module)

    def test_migration_0006_moves_movie_settings(self):
        self.assertIn("0006_notification_settings_consolidation", self.registry)
        self.assertIn(
            "ADD COLUMN IF NOT EXISTS movie_released BOOLEAN NOT NULL DEFAULT TRUE",
            self.registry,
        )
        self.assertIn(
            "ADD COLUMN IF NOT EXISTS movie_release_updates BOOLEAN NOT NULL DEFAULT TRUE",
            self.registry,
        )
        self.assertIn(
            "FROM tv_tracker_final_notification_settings AS final_settings",
            self.registry,
        )
        self.assertIn("DATABASE_SCHEMA_VERSION = 6", self.versions)

    def test_no_runtime_monkey_patching_between_notification_modules(self):
        for forbidden in (
            "unittest.mock",
            "patch(",
            "setattr",
            "view_functions",
            "_ORIGINAL_",
            "_schema_already_prepared",
            "final.ensure_final_schema =",
        ):
            self.assertNotIn(forbidden, self.runtime)
        self.assertIn(
            "final.ensure_final_schema(connection_factory)",
            self.runtime,
        )
        self.assertIn(
            "final.run_final_notification_worker(",
            self.runtime,
        )
        self.assertEqual(
            self.runtime.count("run_final_notification_worker_hardened"),
            1,
        )

    def test_removed_shims_are_gone_from_canonical_owner(self):
        for removed in (
            "serialize_combined_settings",
            "update_combined_settings",
            "list_notifications_final",
            "read_final_settings",
            "FINAL_SETTING_MAP",
            "harden_push_config",
            "tv_tracker_final_notification_settings",
        ):
            self.assertNotIn(removed, self.canonical)

    def test_single_endpoint_registration_per_route(self):
        self.routes = read_source("tvtracker/web/routes.py")
        combined = self.routes + self.canonical
        app_owned = (
            '@app.get("/api/notifications/status")',
            '@app.get("/api/notifications")',
            '@app.post("/api/notifications/read-all")',
            '@app.post("/api/notifications/<int:notification_id>/read")',
            '@app.delete("/api/notifications/<int:notification_id>")',
            '@app.get("/api/notifications/settings")',
            '@app.patch("/api/notifications/settings")',
        )
        canonical_owned = (
            '@app.get("/api/push/config")',
            '@app.get("/api/push/device")',
            '@app.post("/api/push/subscribe")',
            '@app.post("/api/push/unsubscribe")',
            '@app.post("/api/push/presence")',
        )
        for decorator in app_owned:
            self.assertEqual(combined.count(decorator), 1, decorator)
            self.assertEqual(self.routes.count(decorator), 1, decorator)
            self.assertNotIn(decorator, self.canonical)
        for decorator in canonical_owned:
            self.assertEqual(combined.count(decorator), 1, decorator)
            self.assertEqual(self.canonical.count(decorator), 1, decorator)
            self.assertNotIn(decorator, self.routes)

    def test_installer_never_overwrites_notification_endpoints(self):
        for endpoint in (
            "notification_settings_api",
            "notification_settings_patch_api",
            "notifications_api",
            "push_config_api",
            "push_device_api",
            "push_subscribe_api",
            "push_unsubscribe_api",
            "push_presence_api",
        ):
            self.assertNotIn(
                f'app.view_functions["{endpoint}"] =',
                self.canonical,
            )
        self.assertIn('app.view_functions["logout"]', self.canonical)
        self.assertIn('app.view_functions["update_admin_account"]', self.canonical)

    def test_movies_and_tv_flow_through_same_canonical_runner(self):
        self.assertIn("def run_final_notification_worker", self.canonical)
        self.assertIn(
            "movie_result = run_movie_notification_check(connection_factory, tmdb_fetcher, now)",
            self.canonical,
        )
        self.assertIn("core_result = core_runner(now)", self.canonical)
        self.assertIn(
            "return final.run_final_notification_worker(",
            self.runtime,
        )

    def test_notification_persists_before_push_delivery(self):
        ordered = (
            "before = _notification_versions(connection_factory)",
            "core_result = core_runner(now)",
            "movie_result = run_movie_notification_check(connection_factory, tmdb_fetcher, now)",
            "changed = _changed_notifications(connection_factory, before)",
            "queued = enqueue_push_deliveries(connection_factory, changed)",
            "_prepare_push_outbox_state(connection_factory)",
            "push_result = deliver_push_outbox(connection_factory, now)",
        )
        positions = [self.canonical.index(needle) for needle in ordered]
        self.assertEqual(positions, sorted(positions))

    def test_push_is_optional(self):
        self.assertIn('if not config["configured"]:', self.canonical)
        self.assertIn(
            'return {"configured": False, "delivered": 0, "failed": 0, "dead": 0}',
            self.canonical,
        )
        self.assertIn("from pywebpush import WebPushException, webpush", self.canonical)
        self.assertIn("except ImportError:", self.canonical)
        self.assertIn('find_spec("pywebpush")', self.canonical)

    def test_wsgi_installs_polish_then_final_once(self):
        self.assertIn("prepare_final_notification_runtime(database_connection)", self.wsgi)
        self.assertIn("install_notification_polish(app, notifications_module)", self.wsgi)
        self.assertIn("notifications_module.install_final_notifications(", self.wsgi)
        self.assertEqual(self.wsgi.count("install_final_notifications("), 1)
        self.assertIn("run_final_notification_worker_hardened", self.worker)


if __name__ == "__main__":
    unittest.main()

File: tests/test_phase16_notification_consolidation.js
const fs=require("fs"),path=require("path"),assert=require("assert");
const ROOT=path.resolve(__dirname,"..");
const t=fs.readFileSync(path.join(ROOT,"templates/index.html"),"utf8");
const r=fs.readFileSync(path.join(ROOT,"static/js/notifications-runtime.js"),"utf8");
const runtime=fs.readFileSync(path.join(ROOT,"tvtracker/notifications/runtime.py"),"utf8");
const backend=fs.readFileSync(path.join(ROOT,"tvtracker/notifications/push_and_movies.py"),"utf8");
const settingsModule=fs.readFileSync(path.join(ROOT,"tvtracker/notifications/backend.py"),"utf8");
const validation=fs.readFileSync(path.join(ROOT,"tvtracker/notifications/push_validation.py"),"utf8");

function occurrences(value,needle){
    return value.split(needle).length - 1;
}

assert(t.includes("notifications-runtime.js"));
assert(!t.includes("notifications-final.js"));
assert(!t.includes("notifications-polish.js"));
assert(!t.includes("js/notifications.js"));
assert(r.includes("Movie Released"));
assert(r.includes("Push Notifications"));

// The legacy 6-family settings-page renderer is dead code: it is not referenced
// anywhere else and its entry points were removed.
assert(!r.includes("async function renderNotificationSettingsPage"),"the legacy settings-page renderer must be absent");
assert(!r.includes("const SETTINGS_OPTIONS = ["),"the legacy 6-family option list must be absent");
assert(!r.includes("notification-settings-content"),"the legacy settings page container must be absent");
assert(!r.includes("notification-settings-page"),"the legacy settings page id must be absent");

// Exactly one canonical notification settings renderer owns all rows.
assert.strictEqual(occurrences(r,"async function renderNotificationControls"),1,"there must be exactly one canonical settings renderer");
const canonicalStart = r.indexOf("async function renderNotificationControls");
const canonicalEnd = r.indexOf("function openDedicatedSettingsPage",canonicalStart);
assert(canonicalStart > 0 && canonicalEnd > canonicalStart);
const canonical = r.slice(canonicalStart,canonicalEnd);
assert(canonical.includes("list.appendChild(pushRow);"),"the Push control must come from the canonical renderer");
assert(canonical.includes("list.appendChild(masterRow);"),"the master toggle must come from the canonical renderer");
assert(canonical.includes("BASE_SETTING_OPTIONS.forEach"),"TV and movie family rows must come from the canonical renderer");
assert(canonical.includes("pushNotifications"),"the Push row must come from the canonical renderer");
const settingsOwner = r.slice(r.indexOf('const CANONICAL_SETTINGS_ROUTE = "/app/settings/notifications";'));
assert(settingsOwner.includes('["movieReleased","Movie Released"'),"movie settings rows must come from the canonical owner");
assert(settingsOwner.includes('["movieReleaseUpdates","Movie Release Updates"'),"movie release-update rows must come from the canonical owner");

// The Python runtime wrapper is a thin facade: no patching, no duplicated orchestration.
assert(!runtime.includes("_ORIGINAL_ENSURE_FINAL_SCHEMA"),"the runtime wrapper must not keep patch references");
assert(!runtime.includes("_schema_already_prepared"),"the runtime wrapper must not install no-op schema helpers");
assert(!runtime.includes("final.ensure_final_schema ="),"the runtime wrapper must not patch the canonical owner");
assert(runtime.includes("final.ensure_final_schema(connection_factory)"),"runtime preparation must delegate to the canonical schema guard");
assert(runtime.includes("final.run_final_notification_worker("),"the hardened worker must delegate to the canonical worker");
assert(occurrences(runtime,"run_final_notification_worker_hardened") === 1,"there must be exactly one hardened-worker entry point");

// The canonical owner holds every second-audit safeguard natively.
assert(backend.includes("def _prepare_push_outbox_state"),"the outbox preflight must live in the canonical owner");
assert(backend.includes("def run_final_notification_worker"),"the worker orchestration must live in the canonical owner");
assert(backend.includes("DELETE FROM tv_tracker_movie_notification_baseline"),"region-clear must live in the canonical movie runner");
assert(backend.includes("def push_config"),"the VAPID-hardened push config must live in the canonical owner");
assert(backend.includes("validate_vapid_configuration("),"the canonical push config must validate VAPID natively");
assert(!backend.includes("app.view_functions[\"notification_settings_api\"]"),"the installer must not overwrite the settings GET endpoint");
assert(!backend.includes("app.view_functions[\"notification_settings_patch_api\"]"),"the installer must not overwrite the settings PATCH endpoint");
assert(!backend.includes("app.view_functions[\"notifications_api\"]"),"the installer must not overwrite the notifications endpoint");
assert(!backend.includes("tv_tracker_final_notification_settings"),"application code must not read the retired settings table");
assert(!backend.includes("serialize_combined_settings"),"the combined-settings shim must be gone");
assert(!backend.includes("update_combined_settings"),"the combined-settings shim must be gone");
assert(!backend.includes("list_notifications_final"),"the duplicate notifications serializer must be gone");
assert(!validation.includes("harden_push_config"),"monkey-patching of push_config must be gone");
assert(settingsModule.includes('"movieReleased": "movie_released"'),"movie settings must be served by the canonical settings module");
assert(settingsModule.includes('"movieReleaseUpdates": "movie_release_updates"'),"movie settings must be served by the canonical settings module");
assert(settingsModule.includes('"movieId": media_id if media_type == "movie"'),"media-aware notifications must come from the canonical module");
assert(settingsModule.includes('f"/app/movie/{media_id}"'),"movie notifications must route to the movie page");

console.log("Phase 16 notification consolidation contracts passed.");
